reframe/frontend/executors/policies.py

```diff
@@ -221,337 +235,404 @@ def exit(self):
+                return 1
+
         elif self.deps_succeeded(task):
             try:
                 task.setup(task.testcase.partition,
                            task.testcase.environ,
                            sched_flex_alloc_nodes=self.sched_flex_alloc_nodes,
                            sched_options=self.sched_options)
             except TaskExit:
-                return False
+                self._current_tasks.remove(task)
+                return 1
+
+            if isinstance(task.check, RunOnlyRegressionTest):
+                try:
+                    task.compile()
+                    task.compile_wait()
+                except TaskExit:
+                    # Run and run_wait are no-ops for
+                    # CompileOnlyRegressionTest. This shouldn't fail.
+                    self._current_tasks.remove(task)
+                    return 1
+
+                task.policy_stage = 'ready_to_run'
             else:
-                return True
+                task.policy_stage = 'ready_to_compile'
+
+            return 1
+
         elif self.deps_failed(task):
             exc = TaskDependencyError('dependencies failed')
             task.fail((type(exc), exc, None))
-            return False
+            self._current_tasks.remove(task)
+            return 1
         else:
             # Not all dependencies have finished yet
-            return False
+            return 0
 
-    def runcase(self, case):
-        super().runcase(case)
-        check, partition, environ = case
-        self._partitions.add(partition)
+    def advance_ready_to_compile(self, task):
+        if task.check.local or task.check.build_locally:
+            if len(self._local_scheduler_tasks) < self._rfm_max_jobs:
+                try:
+                    task.compile()
+                    task.policy_stage = 'compiling'
+                    self._local_scheduler_tasks.add(task)
+                except TaskExit:
+                    self._current_tasks.remove(task)
 
-        # Set partition-based counters, if not set already
-        self._running_tasks.setdefault(partition.fullname, [])
-        self._ready_tasks.setdefault(partition.fullname, [])
-        self._max_jobs.setdefault(partition.fullname, partition.max_jobs)
+                return 1
+            else:
+                return 0
 
-        task = RegressionTask(case, self.task_listeners)
-        self._task_index[case] = task
-        self.stats.add_task(task)
-        self.printer.status(
-            'RUN', '%s on %s using %s' %
-            (check.name, partition.fullname, environ.name)
-        )
+        partname = task.check.current_partition.fullname
+        if len(self._scheduler_tasks[partname]) < self._max_jobs[partname]:
+            try:
+                task.compile()
+                task.policy_stage = 'compiling'
+                self._scheduler_tasks[partname].add(task)
+            except TaskExit:
+                self._current_tasks.remove(task)
```

Comment:
This is almost identical to the above part. Couldn't we have a specially named pseudo-partition (just for the data structures of the policies) so that we avoid this duplication?

Also is there a reason that you bump only the local tests and not both in case you can?

Reply:
> Also is there a reason that you bump only the local tests and not both in case you can?

I don't understand that comment, I am always trying to bump the test first but I am checking the appropriate job limit.

> Couldn't we have a specially named pseudo-partition (just for the data structures of the policies) so that we avoid this duplication?

This is a bit silly but I wanted to avoid finding a name that will not clash without a real partition. I guess we can put two underscores in the beginning and be safe.

Reply:
> I don't understand that comment, I am always trying to bump the test first but I am checking the appropriate job limit.

I mean that if you have local tests waiting, you will only bump those and return. Eventually, you will bump them all, but wouldn't it make sense to bump both a local and a normal test at once? Just asking.

> This is a bit silly but I wanted to avoid finding a name that will not clash with a real partition. I guess we can put 1 or 2 underscores in the beginning and be safe.

I think the `_rfm` namespace is de facto reserved for ReFrame :-D So `_rfm_local` should be ok.